<a href="https://colab.research.google.com/github/chemoinformatics-lecture/lecture-beginner/blob/main/lesson05_MOPAC/%E3%80%9003%E3%80%91mopac%E3%81%AE%E8%A8%88%E7%AE%97%E3%83%95%E3%82%A1%E3%82%A4%E3%83%AB%E3%81%AE%E7%A2%BA%E8%AA%8D_colab.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# google driveへの接続

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# outファイルの読み取り

In [3]:
with open('/content/drive/MyDrive/data/mopac/AGBQKNBQESQNJD-SSDOTTSWSA-N.out', mode='r') as f:
    lines = f.readlines()
    for line in lines:
        print(line, end='')

 *******************************************************************************
 **                                                                           **
 **                              MOPAC v22.0.4                                **
 **                                                                           **
 *******************************************************************************
 **          Digital Object Identifier (DOI): 10.5281/zenodo.6511958          **
 **    Visit the DOI location for information on how to cite this program.    **
 *******************************************************************************

                              PM7 CALCULATION RESULTS

 *******************************************************************************
 *  CALCULATION DONE:                                Tue Oct 11 07:34:54 2022  *
 *  PM7        - The PM7 Hamiltonian to be used
 *  EF         - USE EF ROUTINE FOR MINIMUM SEARCH
 *  T=         - A TIME OF 172800.0

# arcファイルの読み取り

In [4]:
with open('/content/drive/MyDrive/data/mopac/AGBQKNBQESQNJD-SSDOTTSWSA-N.arc', mode='r') as f:
    lines = f.readlines()
    for line in lines:
        print(line, end='')



                     SUMMARY OF  PM7 CALCULATION

                                                       MOPAC v22.0.4 Linux
                                                       Tue Oct 11 07:34:54 2022

           Empirical Formula: C6 H6 O  =    13 atoms

 PM7 EF
 c1ccc(cc1)O
 InChI=1S/C6H6O/c7-6-4-2-1-3-5-6/h1-5,7H


     GEOMETRY OPTIMISED USING EIGENVECTOR FOLLOWING (EF).     
     SCF FIELD WAS ACHIEVED                                   

          HEAT OF FORMATION       =        -22.15794 KCAL/MOL =     -92.70883 KJ/MOL
          GRADIENT NORM           =          0.72507          =       0.20110 PER ATOM
          DIPOLE                  =          1.30756 DEBYE   POINT GROUP:  Cs  
          NO. OF FILLED LEVELS    =         18
          IONIZATION POTENTIAL    =          9.235839 EV
          HOMO LUMO ENERGIES (EV) =         -9.236  0.035
          MOLECULAR WEIGHT        =         94.1128
          COSMO AREA              =        129.57 SQUARE ANGSTROMS
          COSM

それぞれの意味

HEAT OF FORMATION       =        -22.15781 KCAL/MOL =     -92.70830 KJ/MOL
GRADIENT NORM           =          0.99656          =       0.27640 PER ATOM
DIPOLE                  =          1.30764 DEBYE   POINT GROUP:  Cs  
NO. OF FILLED LEVELS    =         18
IONIZATION POTENTIAL    =          9.235653 EV
HOMO LUMO ENERGIES (EV) =         -9.236  0.035
MOLECULAR WEIGHT        =         94.1128
COSMO AREA              =        129.57 SQUARE ANGSTROMS
COSMO VOLUME            =        119.27 CUBIC ANGSTROMS

# ファイルの読み取り方


In [5]:
# 使用するライブラリーの読み込み
import os
import glob
import pandas as pd

## 1つのarcファイルの読み取り

In [27]:
file_path = "/content/drive/MyDrive/data/mopac/AGBQKNBQESQNJD-SSDOTTSWSA-N.arc"
name = os.path.splitext(os.path.basename(file_path))[0]
print(name)

with open(file_path, mode='r') as f:
    lines = f.readlines()
    #stripメソッドで空白や特定文字の削除
    lines_strip = [line.strip() for line in lines]
    print(lines[10].replace('\n', '')) # 11行目を取得してプリント
    print(lines[11].replace('\n', '')) # 12行目を取得してプリント
    #特定の文字列があるかを判定
    HOMO_line = [line for line in lines_strip if 'HOMO' in line]
    if HOMO_line != []:
      HOMO=float(str(HOMO_line[0][30:39]))
      print("HOMO: " + str(HOMO))

AGBQKNBQESQNJD-SSDOTTSWSA-N
 c1ccc(cc1)O
 InChI=1S/C6H6O/c7-6-4-2-1-3-5-6/h1-5,7H
HOMO: -9.23


## 複数のarcファイルの読み取り

In [44]:
InChIKey = []
inchi = []
smiles = []
HOMO = []

folder_path = "/content/drive/MyDrive/data/mopac/"

for file_path in glob.glob(folder_path + "*.arc"):
  InChIKey.append(os.path.splitext(os.path.basename(file_path))[0])
  print(file_path)
  
  with open(file_path) as f:
    lines = f.readlines()

  lines_strip = [line.strip() for line in lines]

  print(lines[10].replace('\n', ''))
  smiles.append(lines[10].replace('\n', ''))
  inchi.append(lines[11].replace('\n', ''))
    
  HOMO_line = [line for line in lines_strip if 'HOMO' in line]
  if HOMO_line != []:
    HOMO.append(float(str(HOMO_line[0][30:39])))

/content/drive/MyDrive/data/mopac/KKEYFWRCBNTPAC-UHFFFAOYSA-N.arc
 c1ccc(cc1)O
/content/drive/MyDrive/data/mopac/CMCWWLVWPDLCRM-UHFFFAOYSA-N.arc
 c1ccc(cc1)O
/content/drive/MyDrive/data/mopac/OUBMGJOQLXMSNT-UHFFFAOYSA-N.arc
 c1ccc(cc1)O
/content/drive/MyDrive/data/mopac/VRKQEIXDEZVPSY-UHFFFAOYSA-N.arc
 c1ccc(cc1)O
/content/drive/MyDrive/data/mopac/GRFNBEZIAWKNCO-UHFFFAOYSA-N.arc
 c1ccc(cc1)O
/content/drive/MyDrive/data/mopac/NLZUEZXRPGMBCV-UHFFFAOYSA-N.arc
 c1ccc(cc1)O
/content/drive/MyDrive/data/mopac/ZBCATMYQYDCTIZ-UHFFFAOYSA-N.arc
 c1ccc(cc1)O
/content/drive/MyDrive/data/mopac/SEBFKMXJBCUCAI-HKTJVKLFSA-N.arc
 c1ccc(cc1)O
/content/drive/MyDrive/data/mopac/DZLNHFMRPBPULJ-GSVOUGTGSA-N.arc
 c1ccc(cc1)O
/content/drive/MyDrive/data/mopac/GLEVLJDDWXEYCO-AWEZNQCLSA-N.arc
 c1ccc(cc1)O
/content/drive/MyDrive/data/mopac/NWVVVBRKAWDGAB-UHFFFAOYSA-N.arc
 c1ccc(cc1)O
/content/drive/MyDrive/data/mopac/KYAKGJDISSNVPZ-UHFFFAOYSA-N.arc
 c1ccc(cc1)O
/content/drive/MyDrive/data/mopac/YCCILVSKPBXVIP-UHF

In [39]:
f

<_io.TextIOWrapper name='/content/drive/MyDrive/data/mopac/RZFHLOLGZPDCHJ-XZXLULOTSA-N.arc' mode='r' encoding='UTF-8'>

In [41]:
file_path

'/content/drive/MyDrive/data/mopac/RZFHLOLGZPDCHJ-XZXLULOTSA-N.arc'

In [40]:
lines 

['\n',
 '\n',
 '                     SUMMARY OF  PM7 CALCULATION\n',
 '\n',
 '                                                       MOPAC v22.0.4 Linux\n',
 '                                                       Tue Oct 11 07:35:10 2022\n',
 '\n',
 '           Empirical Formula: C6 H6 O  =    13 atoms\n',
 '\n',
 ' PM7 EF\n',
 ' c1ccc(cc1)O\n',
 ' InChI=1S/C6H6O/c7-6-4-2-1-3-5-6/h1-5,7H\n',
 '\n',
 '\n',
 '     GEOMETRY OPTIMISED USING EIGENVECTOR FOLLOWING (EF).     \n',
 '     SCF FIELD WAS ACHIEVED                                   \n',
 '\n',
 '          HEAT OF FORMATION       =        -22.15794 KCAL/MOL =     -92.70883 KJ/MOL\n',
 '          GRADIENT NORM           =          0.72507          =       0.20110 PER ATOM\n',
 '          DIPOLE                  =          1.30756 DEBYE   POINT GROUP:  Cs  \n',
 '          NO. OF FILLED LEVELS    =         18\n',
 '          IONIZATION POTENTIAL    =          9.235839 EV\n',
 '          HOMO LUMO ENERGIES (EV) =         -9.236  0.035

In [33]:
len(InChIKey)

121

In [34]:
InChIKey

['KKEYFWRCBNTPAC-UHFFFAOYSA-N',
 'CMCWWLVWPDLCRM-UHFFFAOYSA-N',
 'OUBMGJOQLXMSNT-UHFFFAOYSA-N',
 'VRKQEIXDEZVPSY-UHFFFAOYSA-N',
 'GRFNBEZIAWKNCO-UHFFFAOYSA-N',
 'NLZUEZXRPGMBCV-UHFFFAOYSA-N',
 'ZBCATMYQYDCTIZ-UHFFFAOYSA-N',
 'SEBFKMXJBCUCAI-HKTJVKLFSA-N',
 'DZLNHFMRPBPULJ-GSVOUGTGSA-N',
 'GLEVLJDDWXEYCO-AWEZNQCLSA-N',
 'NWVVVBRKAWDGAB-UHFFFAOYSA-N',
 'KYAKGJDISSNVPZ-UHFFFAOYSA-N',
 'YCCILVSKPBXVIP-UHFFFAOYSA-N',
 'LYAUICDWKQJAGX-VWLOTQADSA-N',
 'LZJRNLRASBVRRX-ZDUSSCGKSA-N',
 'NAJVRARAUNYNDX-UHFFFAOYSA-N',
 'PWKSKIMOESPYIA-BYPYZUCNSA-N',
 'JDLKFOPOAOFWQN-VIFPVBQESA-N',
 'GVJHHUAWPYXKBD-IEOSBIPESA-N',
 'LSHVYAFMTMFKBA-FPOVZHCZSA-N',
 'UBJVUCKUDDKUJF-UHFFFAOYSA-N',
 'WMBWREPUVVBILR-WIYYLYMNSA-N',
 'QUEDXNHFTDJVIY-RZIGYZOXSA-N',
 'QRYRORQUOLYVBU-JLTOFOAXSA-N',
 'RUZIUYOSRDWYQF-HNNXBMFYSA-N',
 'LCXWQHWWTPOAFI-UWVGGRQHSA-N',
 'UBAXRAHSPKWNCX-UHFFFAOYSA-N',
 'STTGYIUESPWXOW-UHFFFAOYSA-N',
 'BGNXCDMCOKJUMV-UHFFFAOYSA-N',
 'CYJQCYXRNNCURD-AAEUAGOBSA-N',
 'LUSZGTFNYDARNI-UHFFFAOYSA-N',
 'VSWDOR